server/routes/auth.js

```diff
@@ -0,0 +1,336 @@
+                .select('*')
+                .where('username', username)
+                .then(function (row) {
+                    if (row.length > 0) {
+                        row = row[0]
+                        res.status(200);
+                        res.json(row);
+                    } else {
+                        res.status(500).json({ 'status': 'failed', 'error': error });
+                        logger.auth.error('failed to select user')
+                    }
+                })
+                .catch((err) => {
+                    logger.main.error(err);
+                    return next(err);
+                })
+    })
+    .post(isLoggedIn, function (req, res, next) {
+    if (req.body.username === req.user.username) {
+        var username = req.body.username;
+        var givenname = req.body.givenname;
+        var surname = req.body.surname || '';
+        var email = req.body.email;
+        var lastlogondate = Date.now()
+        console.time('insert');
+        db.from('users')
+          .returning('id')
+          .where('username', '=', req.user.username)
+          .update({
+            username: username,
+            givenname: givenname,
+            surname: surname,
+            email: email,
+            lastlogondate: lastlogondate
+          })
+          .then((result) => {
+            console.timeEnd('insert');
+            res.status(200).send({ 'status': 'ok', 'id': result })
+          })
+          .catch((err) => {
+            console.timeEnd('insert');
+            logger.main.error(err)
+            res.status(500).send(err);
+          })
+      } else {
+        res.status(500).json({ message: 'Please update your own details only' });
+        logger.auth.error('Possible attempt to compromise security POST:/auth/profile')
+      }
+    });
+
+router.route('/register')
+    .get(function (req, res, next) {
+        var reg = nconf.get('auth:registration')
+        if (reg === 'enabled') {
+            return res.render('auth', {
+                title: 'Registration',
+                message: req.flash('registerMessage'),
+            });
+        } else {
+            res.redirect('/');
+        }
+    })
+    .post(function (req, res, next) {
+        var reg = nconf.get('auth:registration')
+        if (reg === 'enabled') {
+            const salt = bcrypt.genSaltSync();
+            const hash = bcrypt.hashSync(req.body.password, salt);
+
+            return db('users')
+                .insert({
+                    username: req.body.username,
+                    password: hash,
+                    givenname: req.body.givenname,
+                    surname: req.body.surname,
+                    email: req.body.email,
+                    role: 'user',
+                    status: 'active',
+                    lastlogondate: Date.now()
+                })
+                .then((response) => {
+                    passport.authenticate('login-user', (err, user, info) => {
```

Comment:
Something feels off about this but can't put my finger on it, need to revisit

Reply:
If insert fails softly (i.e. because user already exists) does `.then` execute?

Do all DBs treat inserts strictly? I have a vague recollection that one of the DB engines allowed updates through the insert command

This also doesn't check for conflicts on username does it? Could you register `admin`?

Reply:
Oh man you suck
I'll have to do some tinkering with this to see whats possible.

I was very happy in my bubble why you must pop it.

Reply:
but your correct, the front end does all the validation for dupes. Of you submitted direct to the api you could probably do some shady shit with this one.

Reply:
@davidmckenzie i tested this on MySQL and SQLite, nether seems to give the ability to reset the admin password, MySQL does return a 200 tho, not sure what the go is there. 
I'm still not really comfortable with it tho. 

Any suggestions? Possibly add dupe checking into the route as well?

Reply:
@davidmckenzie put some dupe checking in, can you peruse at your leisure.

Reply:
Much better 👌